spyder/api/widgets/toolbars.py

```diff
@@ -0,0 +1,249 @@
+            self._section_items[section] = [action_or_widget]
+        else:
+            self._section_items[section].append(action_or_widget)
+
+    def _render(self):
+        """
+        Create the toolbar taking into account the sections and locations.
+
+        This method is called once on widget setup.
+        """
+        sec_items = []
+        for sec, items in self._section_items.items():
+            for item in items:
+                sec_items.append([sec, item])
+
+            sep = QAction(self)
+            sep.setSeparator(True)
+            sec_items.append((None, sep))
+
+        if sec_items:
+            sec_items.pop()
+
+        for (sec, item) in sec_items:
+            if isinstance(item, QAction):
+                add_method = super().addAction
+            else:
+                add_method = super().addWidget
+
+            add_method(item)
+
+            if isinstance(item, QAction):
+                text_beside_icon = getattr(item, 'text_beside_icon', False)
+                if text_beside_icon:
+                    widget = self.widgetForAction(item)
+                    widget.setToolButtonStyle(Qt.ToolButtonTextBesideIcon)
+
+                if item.isCheckable():
+                    widget = self.widgetForAction(item)
+                    widget.setCheckable(True)
+
+
+class ApplicationToolBar(SpyderToolBar):
+    """
+    Spyder Main application ToolBar.
+    """
+
+
+class MainWidgetToolbar(SpyderToolBar):
+    """
+    Spyder Widget toolbar class.
+
+    A toolbar used in Spyder Widgets to add internal toolbars
+    to their interface.
+    """
+
+    def __init__(self, parent=None, title=None, location=ToolBarLocation.Top,
+                 corner_widget=None):
+        super().__init__(parent, title=title or '')
+        self._set_corner_widget(corner_widget)
+        self._icon_size = QSize(16, 16)
+
+        # Setup
+        self.setObjectName("main_widget_toolbar_{}".format(
+            str(uuid.uuid4())[:8]))
+        self.setFloatable(False)
+        self.setMovable(False)
+        self.setAllowedAreas(location)
+        self.setContextMenuPolicy(Qt.PreventContextMenu)
+        self.setIconSize(self._icon_size)
+        self._setup_style()
+
+    def set_icon_size(self, icon_size):
+        self._icon_size = icon_size
+        self.setIconSize(icon_size)
+
+    def addWidget(self, widget):
+        """
+        Override Qt method.
+
+        Take into account the existence of a corner widget when adding a new
+        widget in this toolbar.
+        """
+        if self._corner_widget is not None:
+            super().insertWidget(self._corner_separator_action, widget)
+        else:
+            super().addWidget(widget)
+
+    def addAction(self, action):
+        """
+        Override Qt method.
+
+        Take into account the existence of a corner widget when adding a new
+        action in this toolbar.
+        """
+        pass
```

Comment:
Please remove this method if it's unnecessary. Or clarify why you're redefining it to make it pass in its docstring.

Reply:
Fixed